{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Code: Aligning amino acid sequences\n",
    "\n",
    "```{toggle}\n",
    "Here we resuse the code from previous chapters. We have hidden this code to make you focus on the salient parts of the reasoning.\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "tags": [
     "hide-input"
    ]
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "# Print 2 sequences on top of each other\n",
    "def print_alignment(seqA,seqB):\n",
    "    print(seqA)\n",
    "    print(seqB)\n",
    "\n",
    "# Print a dynamic programming score matrix\n",
    "# together with its sequences\n",
    "def print_dynamic(seqA,seqB,dpm):\n",
    "    seqA,seqB = \"-\" + seqA, \"-\" + seqB\n",
    "    m,n = len(seqA),len(seqB)\n",
    "    print('{:^5}'.format(\" \"), end=\"\"),\n",
    "    for j in range(n):\n",
    "        print('{:^5}'.format(seqB[j]), end=\"\")\n",
    "    print()\n",
    "    for i in range(m):\n",
    "        print ('{:^5}'.format(seqA[i]), end=\"\")\n",
    "        for j in range(n):\n",
    "            print ('{:5.1f}'.format(dpm[i,j]), end=\"\")\n",
    "        print()\n",
    "    print()\n",
    "\n",
    "# Format an alignment by inserting gaps in sequences\n",
    "def format_alignment(seqA,seqB,S,trace):\n",
    "    print(\"Best score: \" + str(S[-1,-1]))\n",
    "    outA,outB = \"\",\"\"\n",
    "    i,j = len(seqA),len(seqB)\n",
    "    while i>0 or j>0:\n",
    "        di,dj = trace[i,j]\n",
    "        i += int(di)\n",
    "        j += int(dj)\n",
    "        if di == 0:\n",
    "            outA = \"-\" + outA\n",
    "        else:\n",
    "            outA = seqA[i] + outA\n",
    "        if dj == 0:\n",
    "            outB = \"-\" + outB\n",
    "        else:\n",
    "            outB = seqB[j] + outB\n",
    "    return outA,outB\n",
    "\n",
    "# Initiating dynamic programming matrices, S and trace\n",
    "def initiate_global_dp(m,n):\n",
    "    S = np.zeros((m,n))       # An m*n matrix, initiated with 0's\n",
    "    trace = np.zeros((m,n,2)) # An m*n matrix, initiated with (0,0)'s\n",
    "    S[0,0] = 0.\n",
    "    trace[0,0,:] = (0.,0.)\n",
    "    for i in range(1,m):\n",
    "        S[i,0] = i * gap_penalty()\n",
    "        trace[i,0,:] =(-1,0)\n",
    "    for j in range(1,n):\n",
    "        S[0,j] = j * gap_penalty()\n",
    "        trace[0,j,:] =(0,-1)\n",
    "    return S,trace\n",
    "def global_align(seqA,seqB,print_dynamic_matrix = False):\n",
    "    # Initiating variables\n",
    "    m, n = len(seqA)+1, len(seqB)+1\n",
    "    S,trace = initiate_global_dp(m,n)\n",
    "    # Fill in the rest of the dynamic programming matrix\n",
    "    for i in range(1,m):\n",
    "        for j in range(1,n):\n",
    "            # Note the subtraction of 1 for the sequence position\n",
    "            # In python sequences are indexed from 0 to len-1\n",
    "            match = S[i-1,j-1] + match_score(seqA[i-1],seqB[j-1]) \n",
    "            delete = S[i-1,j] + match_score(seqA[i-1],'-') \n",
    "            insert = S[i,j-1] + match_score('-',seqB[j-1]) \n",
    "            S[i,j] = max(match,delete,insert)\n",
    "            if match >= max(insert,delete):\n",
    "                trace[i,j,:] = (-1,-1)\n",
    "            elif delete >= insert:\n",
    "                trace[i,j,:] = (-1,0)\n",
    "            else:\n",
    "                trace[i,j,:] = (0,-1)\n",
    "    if print_dynamic_matrix:\n",
    "        print_dynamic(seqA,seqB,S)\n",
    "    return format_alignment(seqA,seqB,S,trace)\n",
    "\n",
    "def format_local_alignment(seqA,seqB,S,trace):\n",
    "    outA,outB = \"\",\"\"\n",
    "    i,j = np.unravel_index(S.argmax(),S.shape)\n",
    "    print(\"Best score: \" + str(S[i,j]))\n",
    "    while min(trace[i,j])<0:\n",
    "        di,dj = trace[i,j]\n",
    "        i += int(di)\n",
    "        j += int(dj)\n",
    "        if di == 0:\n",
    "            outA = \"-\" + outA\n",
    "        else:\n",
    "            outA = seqA[i] + outA\n",
    "        if dj == 0:\n",
    "            outB = \"-\" + outB\n",
    "        else:\n",
    "            outB = seqB[j] + outB\n",
    "    return outA,outB\n",
    "\n",
    "# Initiating dynamic programming matrices, S and trace\n",
    "def initiate_local_dp(m,n):\n",
    "    S = np.zeros((m,n))\n",
    "    trace = np.zeros((m,n,2))\n",
    "    S[0,0] = 0.\n",
    "    trace[0,0,:] = (0.,0.)\n",
    "    for i in range(1,m):\n",
    "        S[i,0] = 0\n",
    "        trace[i,0,:] =(0,0)\n",
    "    for j in range(1,n):\n",
    "        S[0,j] = 0\n",
    "        trace[0,j,:] =(0,0)\n",
    "    return S,trace\n",
    "def local_align(seqA,seqB,print_dynamic_matrix = False):\n",
    "    # Initiating variables\n",
    "    m, n = len(seqA)+1, len(seqB)+1\n",
    "    S,trace = initiate_local_dp(m,n)\n",
    "    # Fill in the rest of the dynamic programming matrix\n",
    "    for i in range(1,m):\n",
    "        for j in range(1,n):\n",
    "            match = S[i-1][j-1] + match_score(seqA[i-1],seqB[j-1])\n",
    "            delete = S[i-1,j] + match_score(seqA[i-1],'-') \n",
    "            insert = S[i,j-1] + match_score('-',seqB[j-1]) \n",
    "            S[i,j] = max(match, delete, insert, 0.)\n",
    "            if match >= max(delete,insert,0.):\n",
    "                trace[i,j,:] = (-1,-1.)\n",
    "            elif delete >= max(insert,0.):\n",
    "                trace[i,j,:] = (-1.,0.)\n",
    "            elif insert >= 0.:\n",
    "                trace[i,j,:] = (0.,-1.)\n",
    "            else:\n",
    "                trace[i,j,:] = (0.,0.)\n",
    "    if print_dynamic_matrix:\n",
    "        print_dynamic(seqA,seqB,S)\n",
    "    return format_local_alignment(seqA,seqB,S,trace)\n",
    "\n",
    "def format_semiglobal_alignment(seqA,seqB,S,trace):\n",
    "    outA,outB = \"\",\"\"\n",
    "    m,n = S.shape[0]-1, S.shape[1]-1\n",
    "    i,j = S[:,n].argmax(), S[m,:].argmax()\n",
    "    if S[i,n] > S[m,j]:\n",
    "        print(\"Best score: \" + str(S[i,n]))\n",
    "        j = n\n",
    "        # point the alignmnént from (m,n) to here\n",
    "        for ix in range(i+1,m+1):\n",
    "            print(ix,n)\n",
    "            trace[ix,n,:] = (-1,0)\n",
    "    else:\n",
    "        print(\"Best score: \" + str(S[m,j]))\n",
    "        i = m    \n",
    "        # point the alignmnént from (m,n) to here\n",
    "        for ix in range(j+1,n+1):\n",
    "            print(m,ix)\n",
    "            trace[m,ix,:] = (0,-1)\n",
    "    i,j = m,n\n",
    "    while min(trace[i,j])<0:\n",
    "        di,dj = trace[i,j]\n",
    "        i += int(di)\n",
    "        j += int(dj)\n",
    "        if di == 0:\n",
    "            outA = \"-\" + outA\n",
    "        else:\n",
    "            outA = seqA[i] + outA\n",
    "        if dj == 0:\n",
    "            outB = \"-\" + outB\n",
    "        else:\n",
    "            outB = seqB[j] + outB\n",
    "    return outA,outB\n",
    "\n",
    "# Initiating dynamic programming matrices, S and trace\n",
    "def initiate_semiglobal_dp(m,n):\n",
    "    S = np.zeros((m,n))\n",
    "    trace = np.zeros((m,n,2))\n",
    "    S[0,0] = 0.\n",
    "    trace[0,0,:] = (0.,0.)\n",
    "    for i in range(1,m):\n",
    "        S[i,0] = 0\n",
    "        trace[i,0,:] =(-1,0)\n",
    "    for j in range(1,n):\n",
    "        S[0,j] = 0\n",
    "        trace[0,j,:] =(0,-1)\n",
    "    return S,trace\n",
    "def semiglobal_align(seqA,seqB,print_dynamic_matrix = False):\n",
    "    # Initiating variables\n",
    "    m, n = len(seqA)+1, len(seqB)+1\n",
    "    S,trace = initiate_semiglobal_dp(m,n)\n",
    "    # Fill in the rest of the dynamic programming matrix\n",
    "    for i in range(1,m):\n",
    "        for j in range(1,n):\n",
    "            match = S[i-1][j-1] + match_score(seqA[i-1],seqB[j-1])\n",
    "            delete = S[i-1,j] + match_score(seqA[i-1],'-') \n",
    "            insert = S[i,j-1] + match_score('-',seqB[j-1]) \n",
    "            S[i,j] = max(match, delete, insert)\n",
    "            if match >= max(delete,insert):\n",
    "                trace[i,j,:] = (-1,-1.)\n",
    "            elif delete >= insert:\n",
    "                trace[i,j,:] = (-1.,0.)\n",
    "            else:\n",
    "                trace[i,j,:] = (0.,-1.)\n",
    "    if print_dynamic_matrix:\n",
    "        print_dynamic(seqA,seqB,S)\n",
    "    return format_semiglobal_alignment(seqA,seqB,S,trace)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Protein sequences\n",
    "\n",
    "We can use the exact same alignment methods for protein sequences, as long as we define an appropriate scoring function. Here we will use a PAM250 matrix for the matches, and a gap penalty of 10. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "PAM250 = {\n",
    "'A': {'A': 2,  'C': -2, 'D':  0, 'E': 0, 'F': -3, 'G':  1, 'H': -1, 'I': -1, 'K': -1, 'L': -2, 'M': -1, 'N':  0, 'P':  1, 'Q':  0, 'R': -2, 'S':  1, 'T':  1, 'V':  0, 'W': -6, 'Y': -3},\n",
    "'C': {'A': -2, 'C': 12, 'D': -5, 'E':-5, 'F': -4, 'G': -3, 'H': -3, 'I': -2, 'K': -5, 'L': -6, 'M': -5, 'N': -4, 'P': -3, 'Q': -5, 'R': -4, 'S':  0, 'T': -2, 'V': -2, 'W': -8, 'Y':  0},\n",
    "'D': {'A': 0,  'C': -5, 'D':  4, 'E': 3, 'F': -6, 'G':  1, 'H':  1, 'I': -2, 'K':  0, 'L': -4, 'M': -3, 'N':  2, 'P': -1, 'Q':  2, 'R': -1, 'S':  0, 'T':  0, 'V': -2, 'W': -7, 'Y': -4},\n",
    "'E': {'A': 0,  'C': -5, 'D':  3, 'E': 4, 'F': -5, 'G':  0, 'H':  1, 'I': -2, 'K':  0, 'L': -3, 'M': -2, 'N':  1, 'P': -1, 'Q':  2, 'R': -1, 'S':  0, 'T':  0, 'V': -2, 'W': -7, 'Y': -4},\n",
    "'F': {'A': -3, 'C': -4, 'D': -6, 'E':-5, 'F':  9, 'G': -5, 'H': -2, 'I':  1, 'K': -5, 'L':  2, 'M':  0, 'N': -3, 'P': -5, 'Q': -5, 'R': -4, 'S': -3, 'T': -3, 'V': -1, 'W':  0, 'Y':  7},\n",
    "'G': {'A': 1,  'C': -3, 'D':  1, 'E': 0, 'F': -5, 'G':  5, 'H': -2, 'I': -3, 'K': -2, 'L': -4, 'M': -3, 'N':  0, 'P':  0, 'Q': -1, 'R': -3, 'S':  1, 'T':  0, 'V': -1, 'W': -7, 'Y': -5},\n",
    "'H': {'A': -1, 'C': -3, 'D':  1, 'E': 1, 'F': -2, 'G': -2, 'H':  6, 'I': -2, 'K':  0, 'L': -2, 'M': -2, 'N':  2, 'P':  0, 'Q':  3, 'R':  2, 'S': -1, 'T': -1, 'V': -2, 'W': -3, 'Y':  0},\n",
    "'I': {'A': -1, 'C': -2, 'D': -2, 'E':-2, 'F':  1, 'G': -3, 'H': -2, 'I':  5, 'K': -2, 'L':  2, 'M':  2, 'N': -2, 'P': -2, 'Q': -2, 'R': -2, 'S': -1, 'T':  0, 'V':  4, 'W': -5, 'Y': -1},\n",
    "'K': {'A': -1, 'C': -5, 'D':  0, 'E': 0, 'F': -5, 'G': -2, 'H':  0, 'I': -2, 'K':  5, 'L': -3, 'M':  0, 'N':  1, 'P': -1, 'Q':  1, 'R':  3, 'S':  0, 'T':  0, 'V': -2, 'W': -3, 'Y': -4},\n",
    "'L': {'A': -2, 'C': -6, 'D': -4, 'E':-3, 'F':  2, 'G': -4, 'H': -2, 'I':  2, 'K': -3, 'L':  6, 'M':  4, 'N': -3, 'P': -3, 'Q': -2, 'R': -3, 'S': -3, 'T': -2, 'V':  2, 'W': -2, 'Y': -1},\n",
    "'M': {'A': -1, 'C': -5, 'D': -3, 'E':-2, 'F':  0, 'G': -3, 'H': -2, 'I':  2, 'K':  0, 'L':  4, 'M':  6, 'N': -2, 'P': -2, 'Q': -1, 'R':  0, 'S': -2, 'T': -1, 'V':  2, 'W': -4, 'Y': -2},\n",
    "'N': {'A': 0,  'C': -4, 'D':  2, 'E': 1, 'F': -3, 'G':  0, 'H':  2, 'I': -2, 'K':  1, 'L': -3, 'M': -2, 'N':  2, 'P':  0, 'Q':  1, 'R':  0, 'S':  1, 'T':  0, 'V': -2, 'W': -4, 'Y': -2},\n",
    "'P': {'A': 1,  'C': -3, 'D': -1, 'E':-1, 'F': -5, 'G':  0, 'H':  0, 'I': -2, 'K': -1, 'L': -3, 'M': -2, 'N':  0, 'P':  6, 'Q':  0, 'R':  0, 'S':  1, 'T':  0, 'V': -1, 'W': -6, 'Y': -5},\n",
    "'Q': {'A': 0,  'C': -5, 'D':  2, 'E': 2, 'F': -5, 'G': -1, 'H':  3, 'I': -2, 'K':  1, 'L': -2, 'M': -1, 'N':  1, 'P':  0, 'Q':  4, 'R':  1, 'S': -1, 'T': -1, 'V': -2, 'W': -5, 'Y': -4},\n",
    "'R': {'A': -2, 'C': -4, 'D': -1, 'E':-1, 'F': -4, 'G': -3, 'H':  2, 'I': -2, 'K':  3, 'L': -3, 'M':  0, 'N':  0, 'P':  0, 'Q':  1, 'R':  6, 'S':  0, 'T': -1, 'V': -2, 'W':  2, 'Y': -4},\n",
    "'S': {'A': 1,  'C':  0, 'D':  0, 'E': 0, 'F': -3, 'G':  1, 'H': -1, 'I': -1, 'K':  0, 'L': -3, 'M': -2, 'N':  1, 'P':  1, 'Q': -1, 'R':  0, 'S':  2, 'T':  1, 'V': -1, 'W': -2, 'Y': -3},\n",
    "'T': {'A': 1,  'C': -2, 'D':  0, 'E': 0, 'F': -3, 'G':  0, 'H': -1, 'I':  0, 'K':  0, 'L': -2, 'M': -1, 'N':  0, 'P':  0, 'Q': -1, 'R': -1, 'S':  1, 'T':  3, 'V':  0, 'W': -5, 'Y': -3},\n",
    "'V': {'A': 0,  'C': -2, 'D': -2, 'E':-2, 'F': -1, 'G': -1, 'H': -2, 'I':  4, 'K': -2, 'L':  2, 'M':  2, 'N': -2, 'P': -1, 'Q': -2, 'R': -2, 'S': -1, 'T':  0, 'V':  4, 'W': -6, 'Y': -2},\n",
    "'W': {'A': -6, 'C': -8, 'D': -7, 'E':-7, 'F':  0, 'G': -7, 'H': -3, 'I': -5, 'K': -3, 'L': -2, 'M': -4, 'N': -4, 'P': -6, 'Q': -5, 'R':  2, 'S': -2, 'T': -5, 'V': -6, 'W': 17, 'Y':  0},\n",
    "'Y': {'A': -3, 'C':  0, 'D': -4, 'E':-4, 'F':  7, 'G': -5, 'H':  0, 'I': -1, 'K': -4, 'L': -1, 'M': -2, 'N': -2, 'P': -5, 'Q': -4, 'R': -4, 'S': -3, 'T': -3, 'V': -2, 'W':  0, 'Y': 10}}\n",
    "\n",
    "def gap_penalty():\n",
    "    return -10.0\n",
    "\n",
    "def match_score(letterA,letterB):\n",
    "    if letterA == '-' or letterB == '-':\n",
    "        return gap_penalty()\n",
    "    else:\n",
    "        return PAM250[letterA][letterB]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       -    I    A    M    P    E    P    P    E    D  \n",
      "  -    0.0-10.0-20.0-30.0-40.0-50.0-60.0-70.0-80.0-90.0\n",
      "  I  -10.0  5.0 -5.0-15.0-25.0-35.0-45.0-55.0-65.0-75.0\n",
      "  A  -20.0 -5.0  7.0 -3.0-13.0-23.0-33.0-43.0-53.0-63.0\n",
      "  M  -30.0-15.0 -3.0 13.0  3.0 -7.0-17.0-27.0-37.0-47.0\n",
      "  A  -40.0-25.0-13.0  3.0 14.0  4.0 -6.0-16.0-26.0-36.0\n",
      "  P  -50.0-35.0-23.0 -7.0  9.0 13.0 10.0  0.0-10.0-20.0\n",
      "  E  -60.0-45.0-33.0-17.0 -1.0 13.0 12.0  9.0  4.0 -6.0\n",
      "  P  -70.0-55.0-43.0-27.0-11.0  3.0 19.0 18.0  8.0  3.0\n",
      "  T  -80.0-65.0-53.0-37.0-21.0 -7.0  9.0 19.0 18.0  8.0\n",
      "  I  -90.0-75.0-63.0-47.0-31.0-17.0 -1.0  9.0 17.0 16.0\n",
      "  D  -100.0-85.0-73.0-57.0-41.0-27.0-11.0 -1.0 12.0 21.0\n",
      "  E  -110.0-95.0-83.0-67.0-51.0-37.0-21.0-11.0  3.0 15.0\n",
      "\n",
      "Best score: 15.0\n",
      "IAMAPEPTIDE\n",
      "IAM-PEPP-ED\n"
     ]
    }
   ],
   "source": [
    "seqA,seqB = global_align(\"IAMAPEPTIDE\",\"IAMPEPPED\",True)\n",
    "print_alignment(seqA,seqB)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Best score: 3.0\n",
      "MYPERFECTDAY\n",
      "T-HER-ESAMAY\n"
     ]
    }
   ],
   "source": [
    "seqA,seqB = global_align(\"MYPERFECTDAY\",\"THERESAMAY\",False)\n",
    "print_alignment(seqA,seqB)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Best score: 14.0\n",
      "PERFECTDAY\n",
      "HER-ESAMAY\n"
     ]
    }
   ],
   "source": [
    "seqA,seqB = local_align(\"MYPERFECTDAY\",\"THERESAMAY\",False)\n",
    "print_alignment(seqA,seqB)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
